### DISCUSSION TITLE
Using objects between different sessions.

### DISCUSSION THREAD

**Discussion by TilmanK** [sqlalchemy/sqlalchemy#7452] [Usage Questions] [answered]
Hey guys,

I've come to a point where I'm absolutely frustrated, and you're probably my last chance of getting me out of this :/.

I'm trying to use SqlAlchemy objects between two sessions, which works like a charm, mostly. So, I load a database object and its relationships into my PyQt application, let the user do some stuff and write all the things back into the database when he hits the save button. Furthermore, I'm also able to detect changes on that object (and it's related objects), which isn't only cool, but saves me from writing a lot of code.

My problem is demonstrated in the code below, that I have boiled down to a minimum example. Here are the table definitions:

```py
class ProjectType(Base):

    __tablename__ = "project_types"

    project_type_id = Column(String(7), primary_key=True)
    name = Column(String)


class Project(Base):

    __tablename__ = "projects"

    id = Column(Integer, primary_key=True)
    project_type_id = Column(String(7), ForeignKey("project_types.id"), nullable=False)
    predecessor_id = Column(Integer, ForeignKey("projects.id"))


    project_type = relationship(ProjectType, uselist=False)
    predecessor = relationship("Project", remote_side=[id], uselist=False)
```

And here's how to produce the problem.
```py

# load the project with its type
session = Session()
project = session.query(Project).filter_by(id=1).options(joinedload(Project.project_type)).one()
session.close()

# do some stuff inbetween, Qt event loop, blabla

# load the predecessor which has the SAME project_type_id
session = Session()
project.predecessor = session.query(Project).filter_by(id=2).options(joinedload(Project.project_type)).one()
session.close()

# now check if changes happened
session = Session()
session.add(project)
print("Changes: " + str(session.is_modified(project)))
session.expunge(project)
session.close()
```

The code breaks at the `session.add` when I try to add the Project instance with its predecessor back into a session and where I get a problem, when both project_type relationships refer to the same row but within different objects as they were loaded in different sessions.

What I get:
```<class 'sqlalchemy.exc.InvalidRequestError'> Can't attach instance <ProjectType at 0x2c0a3a99670>; another instance with key (<class 'src.model.database.ProjectType'>, ('test',), None) is already present in this session.```

That's ok and what to expect from what I read in the docs. I've tried to do something with merge or configure the cascade of that relationship to merge without any success. 

So, what is the correct way to do this?

**Comment by TilmanK** [comment, marked as answer]
The solution is simple, I can do the following when loading the predecessor:

```py
# load the predecessor which has the SAME project_type_id
session = Session()
project.project_type = session.merge(project.project_type, load=False)
project.predecessor = session.query(Project).filter_by(id=2).options(joinedload(Project.project_type)).one()
session.close()
```
Nevertheless, I don't know if it's the right approach.

**Reply by zzzeek** [reply]
it looks like it is.  whenever you have state from other database sessions, the correct way to reconcile it within a new session which might already have that identity present is to use merge().  so if using merge() made it work, then yes that makes total sense.

**Reply by TilmanK** [reply]
Thanks, Mike!
